### ISSUE TITLE
Work on a11y primitives

### ISSUE THREAD

**Issue by IniZio** [IniZio/form-builder#6]
First off, there is no silver bullet to a11y. Making individual components a11y-friendly provides limited value if the website as a whole has major roadblocks for disabled. The best way is always to find disabled people and tune the whole website accordingly. Before having resource to do that though, can probably emulate that yourself with screen-reader e.g. Windows Narrator, Linux Orca, Mac Voice Over.

AT this stage I have very limited understanding to a11y so will still start with making individual components that will probably be useful in practical projects first

Launch Orca: Super + Alt + S

Side note: Apparently firefox doesn't work immediately on Orca launched. This workaround worked ([ref](https://bugzilla.mozilla.org/show_bug.cgi?id=956684))

```
sudo su
cd /usr/lib
ln -s i386-linux-gnu/gtk-2.0 .
```

### Bookmarks

- https://accessible-vue.com
- https://www.w3.org/WAI/ARIA/apg/

### Scribbles

Most of the time, should not have programmatic focus change unless dom structure does not match desired focus order e.g. Disclosure trigger is under its content.

Seems it is suggested to first set `tabindex="-1"` before programmatically focus on the element e.g. the Skip to element on route change

If an icon button, preferably label the icon with the button

A drawer should have a nav. Doesn't have to be the outside-most layer. Also should have focus trap

There are 2 ways to implementing a tab component. One is arrow key navigation and another is just tab everything. I prefer arrow key

Live region: Basically "toast" for for non-visual readers. It does not only mean real toast messages but also events that are not obvious to disabled, e.g. route changes in SPA since it does not involve reloading of the browser

Has eslint plugin for a11y e.g. eslint-plugin-vuejs-accessibility for vue

Focus outline and really useful (`:focus-visible`). for route changes, focusing on the container is most consistent compared to on header

Prefer `aria-label` since quite some screen readers have bug and will read the labels twice

### Components

This will be a list of components I want to build, probably add more along the way of reading more projects/docs/tutorials

- [ ] #10
- [ ] #7 
- [ ] #8
- [ ] #9
- [ ] #11 
- [ ] #12
- [ ] #13
- [ ] #14
- [ ] #15

#### Skip-to

Used when the main content that is not immediately in focus on visit page. For example you see "Skip to main content" in google search page if you press tab, or "Skip to content" in github main page

Maybe the focus on route change configurable since some frameworks already have similar function

Notes:
- It should point to the main content e.g. `<main>` tag. 
- It should be the first focus-able element

#### Disclosure

aka. Accordion

- Trigger has `role="button"`
- Still focus on the trigger on expand/collapse
- Trigger has `aria-expanded="true"`/`aria-expanded="false"` according to expand state
- Hidden content should be hidden i.e. `display: none` / `visibility: hidden`

#### Modal

Interestingly, a Modal actually doesn't necessarily mean a dialog, but rather a part of page that will change the mode of website, blocking it and requiring user interaction. Therefore as long as user has to dismiss the element before moving on to the rest of the website, then it's a modal

example implementation: https://github.com/morkro/vue-a11y-dialog

- Its is highly recommended to have a way of e.g. <kbd>ESC</kbd> to exit dialog, close button
- Although there is a `aria-modal`, it is pretty broken so better not use it just yet. Instead 

#### Menu

https://marcus.io/blog/menu-or-not
